APPS-1421 Update dir output for WDL 2.0

Gvaihir opened the pull request:
This PR is dependent on https://github.com/dnanexus/wdlTools/pull/235

Gvaihir committed:
APPS-1421 Directory output
(fix) build.sbt - dependency

Gvaihir committed:
Merge branch 'develop' into APPS-1421_dir_output

# Conflicts:
#	build.sbt

Gvaihir committed:
APPS-1421 dir outputs
(fix) Type.scala - Directory coerces to non-primitive
(fix) ValueSerde.scala - VFolder does not depend on paths as objects

Gvaihir committed:
APPS-1421 dir outputs
(fix) ValueSerde.scala - VFile in the VFolder has a complex path. Unit test, integration test.

Gvaihir committed:
APPS-1421 dir outputs
(feat) RELEASE_NOTES.md, ExpertOptions.md - updt

emiloslavsky commented on doc/ExpertOptions.md:
````diff
@@ -232,13 +235,40 @@ is passed to dxCompiler using the `-input` option, it is transformed into the fo
+  }
+}
+```
+The `folderoutput/` directory has to be specified **without** `dx://` prefix or project ID. It has to exist on the worker 
````
is project.contribute access required to run workflows and outputs with output directories?

Gvaihir replied:
I did not introduce any changes to the delocalization logic, neither I did for the way the applets/workflows are compiled. Therefore no specific tests to confirm access permissions were done.

commandlinegirl replied:
@Gvaihir Would you investigate whether such access is needed?

Gvaihir replied:
@commandlinegirl , @emiloslavsky - could you explain how this test would look like? If you have no "Contribute" rights to the project, how can you compile a WF/make an applet there?

commandlinegirl replied:
It's about what project access the workflow's applets require at runtime for this feature to work, please check: https://documentation.dnanexus.com/developer/apps/app-metadata?q=dxapp.json#access-field-app-permissions

Gvaihir replied:
Thanks! But that was my point - I did not change the way execs are compiled. Anyway, tested with extras.json (per @emiloslavsky suggestions). Added to integration tests, [16f8565](https://github.com/dnanexus/dxCompiler/pull/396/commits/16f8565e2b58765982bbbf9d8b1781a188f7abf5). No "CONTRIBUTE" access needed.

commandlinegirl commented on doc/ExpertOptions.md:
```diff
@@ -232,13 +235,40 @@ is passed to dxCompiler using the `-input` option, it is transformed into the fo
 * Enact policies and practices to prevent modification of folders that will be used as input when job reuse is enabled.
 
+###### Directory outputs:
+WDL 2.0 supports directory outputs. Specify directory outputs in your tasks/workflows by providing an output pat as a string.   
```
?

```suggestion
WDL 2.0 supports directory outputs. Specify directory outputs in your tasks/workflows by providing an output part as a string.   
```

Gvaihir replied:
`path` - fixed in [8f535ea](https://github.com/dnanexus/dxCompiler/pull/396/commits/8f535ea1a6f4d1a6bfbb0c819e5f1cdbe20a96b6)

commandlinegirl commented on doc/ExpertOptions.md:
```diff
@@ -210,12 +210,15 @@ which, on the worker, results in a file `foo.txt` being created in the inputs di
 
-Both CWL and the development version of WDL have a `Directory` data type. Although DNAnexus does not treat folders as first-class objects, dxCompiler does support `Directory`-typed inputs and outputs, with some caveats.
+Both CWL and the development version of WDL have a `Directory` data type. Although DNAnexus does not treat folders as 
+first-class objects, dxCompiler does support `Directory`-typed inputs and outputs, with some caveats.
```
I'm wondering what all the caveats are. Is it already specified in the documentation? If not, I think it'd be great to add the info on them.

Gvaihir replied:
they are in the text below. I highlighted the sections with a respective header "Caveats" in [bcdbd92](https://github.com/dnanexus/dxCompiler/pull/396/commits/bcdbd927032d6fd4b2c187e9646bf3e12803d6a3)

Gvaihir committed:
APPS-1421 dir outputs
(feat) RELEASE_NOTES.md, ExpertOptions.md - updt

Gvaihir committed:
APPS-1421 dir outputs
(feat) ExpertOptions.md - updt

Gvaihir committed:
APPS-1421 dir outputs
(feat) apps_1421_dir_output_extras.json - extras

Gvaihir committed:
APPS-1421 dir outputs
(feat) apps_1421_dir_output_extras.json - no access for all stages

commandlinegirl reviewed (dismissed):
Thank you @Gvaihir, this looks good to me! I still have some questions I added to the ticket but it may be out of scope for this task.

Could you add an info to the ExpertOption.md doc that the output in a DNAnexus native applet is represented as a hash type (as a question about this was asked during our morning meeting and couldn't find an info in the docs)? Do you think it’s worthy to modify the test you added or add a test with two jobs a & b, where a’s output dir gets pass to the input dir of b.

Gvaihir commented:
All of the above is in principal in the scope - I'll modify accordingly
Requested changes in [9e9424c](https://github.com/dnanexus/dxCompiler/pull/396/commits/9e9424cd0a414faeefffb89267645d5aa6be6b9c)

Gvaihir committed:
APPS-1421 dir outputs
(feat) apps_1421_dir_output* - addressing reviews
(feat) ExpertOptions.md - Directory is a hash

commandlinegirl commented on test/wdl_2_0/apps_1421_dir_output_extras.json:
```diff
@@ -19,7 +19,7 @@
       "runSpec": {
         "access": {
-          "project": "NONE",
+          "project": "VIEW",
```
This setting is redundant since by default applets have VIEW access to the project
https://documentation.dnanexus.com/developer/api/running-analyses#example-chained-execution
https://documentation.dnanexus.com/developer/api/running-analyses#project-permissions

Gvaihir replied:
Yep, sorry, just carryover from testing different combinations. I'll change if other changes needed. Otherwise don't want to trigger GHA again

Gvaihir committed:
APPS-1421 dir outputs
(feat) ExpertOptions.md - Caveat 3

commandlinegirl reviewed (dismissed):
LGTM, thank you!

Gvaihir committed:
APPS-1421 dir outputs
(fix) run_tests.py - test_name
(fix) apps_1421_dir_output_extras.json - redundant option
